XamarinCommunityToolkit/Effects/SafeAreaEffectRouter.ios.cs

```diff
@@ -24,24 +24,27 @@ protected override void OnAttached()
 
 			var insets = UIApplication.SharedApplication.Windows[0].SafeAreaInsets;
 
-			initialMargin = element.Margin;
-
 			if (insets.Top <= 0)
 				return;
 
+			initialMargin = element.Margin;
+
 			var safeArea = SafeAreaEffect.GetSafeArea(element);
 
 			element.Margin = new Thickness(
-				initialMargin.Left + (safeArea.Left ? insets.Left : 0),
-				initialMargin.Top + (safeArea.Top ? insets.Top : 0),
-				initialMargin.Right + (safeArea.Right ? insets.Right : 0),
-				initialMargin.Bottom + (safeArea.Bottom ? insets.Bottom : 0));
+				element.Margin.Left + (safeArea.Left ? insets.Left : 0),
```

Comment:
Why not use the `initialMargin`? Here we are calling two get methods per call, if we use the `initialMargin` will be just one call.

Reply:
@pictos
There would be the same number of “calls” with initialMargin.Value.Xxx. And even if there was one more “call”, the compiler is pretty good at taking care of micro optimizations like that, not to mention that I’d gladly trade like one cpu cycle for legibility (as in store initial value, then modify value). But in this case the cost is even the same. If you really want to go down to the nitty gritty:

element.Margin.Left vs initialValue.Value.Left should be equivalent
There is no method call involved in any of these, just a property fetcher, that surely will be inlined anyway
I think discussing individual cycles takes away time from more important matters, especially if it’s at the cost of legibility.

Reply:
@Tommigun1980
> element.Margin.Left vs initialValue.Value.Left should be equivalent

Why `initialMargin` is nullable? If I'm not wrong the way that XF works, we can guarantee that `initialMargin` wouldn't be null.

> There is no method call involved in any of these, just a property fetcher, that surely will be inlined anyway

I did a test on my side and this isn't true. Using a local variable to get the `Margin` and access the values will be away faster than using your approach. I ran against .NETCore, Mono should follow the same results but with different values.

``` ini

BenchmarkDotNet=v0.12.1.1356-nightly, OS=Windows 10.0.19041.508 (2004/?/20H1)
Intel Core i5-7300HQ CPU 2.50GHz (Kaby Lake), 1 CPU, 4 logical and 4 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT


```
|                     Method |     Mean |   Error |  StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |---------:|--------:|--------:|------:|--------:|-------:|------:|------:|----------:|
| UsingLocalToAccessProperty | 205.6 ns | 4.13 ns | 7.24 ns |  1.00 |    0.00 | 0.0153 |     - |     - |      48 B |
|     AccessPropertyDirectly | 329.8 ns | 6.42 ns | 6.00 ns |  1.65 |    0.07 | 0.0153 |     - |     - |      48 B |

You can run by yourself [here](https://github.com/pictos/XamarinCommunityToolkit/tree/benchmark). If I did something wrong in my preparation please let me know.

>  if it’s at the cost of legibility.
I'm sorry but I can't see the cost of legibility to using a local variable instead of using a property directly.

Reply:
Every time you call getter of "Element.Margin" the unboxing happens (because Thickness is a struct, but all bindable properties stores "object")

Reply:
> @Tommigun1980
> 
> > element.Margin.Left vs initialValue.Value.Left should be equivalent
> 
> Why `initialMargin` is nullable? If I'm not wrong the way that XF works, we can guarantee that `initialMargin` wouldn't be null when we need to use it. And `Thickness` is a `struct`.
> 
> > There is no method call involved in any of these, just a property fetcher, that surely will be inlined anyway
> 
> I did a test on my side and this isn't true. Using a local variable to get the `Margin` and access the values will be away faster than using your approach. I ran against .NETFramework, Mono should follow the same results but with different values.
> 
> ```ini
> BenchmarkDotNet=v0.12.1.1356-nightly, OS=Windows 10.0.19041.508 (2004/?/20H1)
> Intel Core i5-7300HQ CPU 2.50GHz (Kaby Lake), 1 CPU, 4 logical and 4 physical cores
>   [Host]     : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT
>   DefaultJob : .NET Framework 4.8 (4.8.4220.0), X64 RyuJIT
> ```
> 
> Method	Mean	Error	StdDev	Ratio	RatioSD	Gen 0	Gen 1	Gen 2	Allocated
> UsingLocalToAccessProperty	205.6 ns	4.13 ns	7.24 ns	1.00	0.00	0.0153	-	-	48 B
> AccessPropertyDirectly	329.8 ns	6.42 ns	6.00 ns	1.65	0.07	0.0153	-	-	48 B
> You can run by yourself [here](https://github.com/pictos/XamarinCommunityToolkit/tree/benchmark). If I did something wrong in my preparation please let me know.
> 
> > if it’s at the cost of legibility.
> 
> I'm sorry but I can't see the cost of legibility to using a local variable instead of using a property directly.

Why it’s nullable? What do you mean, the FIX in this PR is to make it nullable. It’s not set in all cases when disabled (old iOS etc).

Also I’m closing this, REALLY don’t have time for these kinds of discussions.

Bye.

Reply:
@Tommigun1980, I'm sorry if I sounded unfriendly, English isn't my main language. I'm doing a code review and I don't have all context even reading the PR description and issue. Said that let me rephrase my concerns.

We're building here a lib that will be used for many developers in many applications, which different contexts, so performance is important to preserve especially if it's an easy win like this.

Making a struct nullable implies adding checks to make sure that we don't face a null exception, this increases complexity to the code, so if we can avoid that is better.

And maybe I'm wrong and I'm ok with that, it's always a chance to learning something new. And please remember this is an open-source project so this kind of discussion can happen in all PRs.